File: zanshin/services/exports.py
"""Export formats for issues: SARIF, OpenVEX and CSV.

VEX was the point of storing triage decisions in the standard's vocabulary rather
than in free text — so this is a serialization, not a translation. Every field an
OpenVEX statement needs (status, justification, timestamps, product identity) is
already on `Issue`; nothing here has to infer or invent anything, which is what
makes the output trustworthy enough to hand to a customer or an auditor.

SARIF is the odd one out in purpose: OpenVEX and CSV are for people outside the
pipeline (customers, auditors, spreadsheets), while SARIF exists so a finding stops
living only in Zanshin. It is what GitHub code scanning, GitLab and Azure DevOps
ingest natively, which is what puts a problem in front of the developer who
introduced it, annotated on the line, in the pull request — instead of on a
dashboard they have no reason to open.

Pure functions over model objects: the HTTP layer decides how to deliver them,
and a UI download button can reuse them unchanged.
"""
import csv
import io
from typing import Any, Dict, Iterable, List, Optional

from zanshin.models.issue import (
    STATE_RESOLVED,
    TRIAGE_AFFECTED,
    TRIAGE_FIXED,
    TRIAGE_NOT_AFFECTED,
    TRIAGE_UNDER_REVIEW,
    Issue,
)

OPENVEX_CONTEXT = "https://openvex.dev/ns/v0.2.0"

# Zanshin's triage vocabulary is already OpenVEX's, with one exception:
# `under_review` is spelled `under_investigation` in the specification.
_VEX_STATUS = {
    TRIAGE_UNDER_REVIEW: "under_investigation",
    TRIAGE_AFFECTED: "affected",
    TRIAGE_NOT_AFFECTED: "not_affected",
    TRIAGE_FIXED: "fixed",
}

CSV_COLUMNS = [
    "id", "type", "identifier", "severity", "cvss_score", "epss_score", "is_kev",
    "package_name", "package_version", "purl", "dependency", "file_path", "line",
    "fix_state", "fix_versions", "state", "triage_status", "triage_justification",
    "triaged_by", "triaged_at", "triage_expires_at", "first_seen_at",
    "last_seen_at", "times_seen", "link",
]

SARIF_VERSION = "2.1.0"
SARIF_SCHEMA = "https://json.schemastore.org/sarif-2.1.0.json"

# SARIF has four levels and no notion of "critical". Anything a security tool
# would call critical or high has to land on "error", because "warning" is what a
# reviewer scrolls past.
_SARIF_LEVEL = {
    "critical": "error",
    "high": "error",
    "medium": "warning",
    "low": "note",
    "negligible": "note",
    "unknown": "warning",
}

# GitHub ranks and filters on this property, not on `level`, so it is what keeps a
# critical distinguishable from a high once both are "error". The numbers follow
# the CVSS bands GitHub documents.
_SECURITY_SEVERITY = {
    "critical": "9.5",
    "high": "8.0",
    "medium": "5.5",
    "low": "3.0",
    "negligible": "1.0",
}

_ISSUE_TYPE_LABEL = {
    "vulnerability": "Vulnérabilité",
    "secret": "Secret exposé",
    "iac": "Configuration d'infrastructure",
    "license": "Licence",
    "ai_review": "Revue IA",
}


def build_openvex_document(
    issues: Iterable[Issue],
    *,
    author: str,
    product_id: str,
    document_id: str,
    timestamp: str,
    version: int = 1,
) -> Dict[str, Any]:
    """An OpenVEX document for one product, from its vulnerability issues.

    Only `type="vulnerability"` issues are included: VEX is defined over
    vulnerability identifiers, and a hardcoded secret or a failed IaC check has no
    CVE to make a statement about. Issues without an identifier are skipped for
    the same reason — an anonymous statement isn't one.

    `timestamp`, `document_id` and `author` are supplied by the caller rather than
    computed here: a VEX document is a signed-ish assertion about who said what
    and when, so those belong to whoever is publishing it, not to a helper.
    """
    statements = []
    for issue in issues:
        if issue.type != "vulnerability" or not issue.identifier:
            continue

        statement: Dict[str, Any] = {
            "vulnerability": {"name": issue.identifier},
            "products": [{"@id": product_id}],
            "status": _VEX_STATUS.get(issue.triage_status, "under_investigation"),
        }
        # A resolved-but-never-triaged issue is factually fixed: the scanner
        # stopped seeing it. Saying "under investigation" about something that is
        # gone would be misleading in a document meant to answer exactly that.
        if issue.state == STATE_RESOLVED and issue.triage_status == TRIAGE_UNDER_REVIEW:
            statement["status"] = "fixed"

        if statement["status"] == "not_affected":
            # Required by the specification for this status, and guaranteed
            # present by `IssueService.triage`.
            statement["justification"] = issue.triage_justification
            if issue.triage_comment:
                statement["impact_statement"] = issue.triage_comment
        elif statement["status"] == "affected" and issue.triage_comment:
            # For "affected", free text belongs in the action statement.
            statement["action_statement"] = issue.triage_comment

        if issue.purl:
            statement["products"] = [{"@id": product_id, "identifiers": {"purl": issue.purl}}]
        if issue.triaged_at:
            statement["timestamp"] = issue.triaged_at.isoformat()
        elif issue.last_seen_at:
            statement["timestamp"] = issue.last_seen_at.isoformat()

        statements.append(statement)

    return {
        "@context": OPENVEX_CONTEXT,
        "@id": document_id,
        "author": author,
        "timestamp": timestamp,
        "version": version,
        "tooling": "Zanshin",
        "statements": statements,
    }


def build_sarif_document(
    issues: Iterable[Issue],
    *,
    target_name: str,
    tool_version: str = "1.0.0",
    information_uri: Optional[str] = None,
) -> Dict[str, Any]:
    """A SARIF 2.1.0 log for one target's issues.

    Design decisions worth stating, because SARIF is permissive enough that a
    technically valid document can still be useless in a code-scanning UI:

    * **Triaged issues are `suppressions`, not omissions.** Dropping them would
      make a platform re-report them as new on the next upload, undoing the triage
      work; and a suppression carries the justification, so the reviewer sees *why*
      it was dismissed. `not_affected` and `fixed` are suppressed, `affected` is
      not — a decision that the problem is real must stay visible.
    * **Resolved issues are excluded entirely.** They are gone; SARIF's job here is
      the current state of the branch being built.
    * **`partialFingerprints` carries Zanshin's own fingerprint**, which is what
      lets the platform match an issue across uploads even when the file moves or
      the line shifts — the same identity that keeps triage attached across scans.
    * **Every result gets a location**, falling back to the repository root when a
      dependency issue has no file. GitHub silently drops results without one, so
      an "honest" empty location would mean the vulnerability findings — the bulk
      of them — never appeared at all.

    Rules are emitted per distinct identifier rather than per issue, because that
    is what SARIF's model means by a rule and what lets a platform group findings.
    """
    issues = [issue for issue in issues if issue.state != STATE_RESOLVED]

    rules: Dict[str, Dict[str, Any]] = {}
    rule_index: Dict[str, int] = {}
    results: List[Dict[str, Any]] = []

    for issue in issues:
        rule_id = _sarif_rule_id(issue)
        if rule_id not in rules:
            rule_index[rule_id] = len(rules)
            rules[rule_id] = _sarif_rule(issue, rule_id)

        result: Dict[str, Any] = {
            "ruleId": rule_id,
            "ruleIndex": rule_index[rule_id],
            "level": _SARIF_LEVEL.get((issue.severity or "unknown").lower(), "warning"),
            "message": {"text": _sarif_message(issue)},
            "locations": [_sarif_location(issue)],
            "partialFingerprints": {"zanshinIssueFingerprint": issue.fingerprint},
            "properties": {
                "zanshinIssueId": issue.id,
                "type": issue.type,
                "firstSeen": _isoformat(issue.first_seen_at),
                "timesSeen": issue.times_seen or 1,
            },
        }
        if issue.is_direct_dependency is not None:
            result["properties"]["dependency"] = (
                "direct" if issue.is_direct_dependency else "transitive"
            )
        if _is_suppressed(issue):
            result["suppressions"] = [{
                # "external": the decision was made in Zanshin, not in an
                # annotation in the source, which is what this kind documents.
                "kind": "external",
                "justification": _suppression_justification(issue),
            }]
        results.append(result)

    return {
        "$schema": SARIF_SCHEMA,
        "version": SARIF_VERSION,
        "runs": [{
            "tool": {"driver": {
                "name": "Zanshin",
                "version": tool_version,
                **({"informationUri": information_uri} if information_uri else {}),
                "rules": list(rules.values()),
            }},
            "results": results,
            "properties": {"target": target_name},
        }],
    }


def _sarif_rule_id(issue: Issue) -> str:
    """Stable, and namespaced by type.

    A gitleaks rule and a checkov check can collide on an identifier, and a
    platform keyed on `ruleId` would then merge two unrelated classes of problem
    under one heading.
    """
    identifier = issue.identifier or "unspecified"
    return f"zanshin/{issue.type}/{identifier}"


def _sarif_rule(issue: Issue, rule_id: str) -> Dict[str, Any]:
    label = _ISSUE_TYPE_LABEL.get(issue.type, issue.type)
    rule: Dict[str, Any] = {
        "id": rule_id,
        "name": (issue.identifier or issue.type).replace(" ", ""),
        "shortDescription": {"text": f"{label} : {issue.identifier or 'non identifié'}"},
        "properties": {"tags": ["security", issue.type]},
    }
    if issue.description:
        rule["fullDescription"] = {"text": issue.description[:1000]}
    if issue.link:
        rule["helpUri"] = issue.link
    severity = (issue.severity or "").lower()
    if severity in _SECURITY_SEVERITY:
        rule["properties"]["security-severity"] = _SECURITY_SEVERITY[severity]
    return rule


def _sarif_message(issue: Issue) -> str:
    """What the developer reads in the pull request, so it says what to do.

    The fixed version is the single most useful thing to put in front of someone
    who has thirty seconds: it turns "there is a CVE" into "change this line".
    """
    parts = []
    if issue.package_name:
        package = issue.package_name
        if issue.package_version:
            package += f" {issue.package_version}"
        parts.append(package)
    parts.append(issue.identifier or _ISSUE_TYPE_LABEL.get(issue.type, issue.type))
    message = " — ".join(parts)

    if issue.fix_versions:
        message += f" — corrigé dans {issue.fix_versions}"
    elif issue.fix_state == "not-fixed":
        message += " — aucun correctif publié"
    if issue.is_kev:
        message += " — exploitation active connue (CISA KEV)"
    if issue.is_direct_dependency is False:
        message += " — dépendance transitive"
    return message


def _sarif_location(issue: Issue) -> Dict[str, Any]:
    location: Dict[str, Any] = {
        "physicalLocation": {
            # A relative URI, as SARIF requires for source that the consumer
            # resolves against the repository it just checked out.
            "artifactLocation": {"uri": issue.file_path or "."},
        }
    }
    if issue.line:
        location["physicalLocation"]["region"] = {"startLine": int(issue.line)}
    if issue.purl:
        location["logicalLocations"] = [{"name": issue.purl, "kind": "package"}]
    return location


def _is_suppressed(issue: Issue) -> bool:
    return issue.triage_status in (TRIAGE_NOT_AFFECTED, TRIAGE_FIXED)


def _suppression_justification(issue: Issue) -> str:
    parts = [issue.triage_status]
    if issue.triage_justification:
        parts.append(issue.triage_justification)
    if issue.triage_comment:
        parts.append(issue.triage_comment)
    if issue.triaged_by:
        parts.append(f"décidé par {issue.triaged_by}")
    if issue.triage_expires_at:
        parts.append(f"à revoir le {issue.triage_expires_at.date().isoformat()}")
    return " — ".join(parts)


def build_issues_csv(issues: Iterable[Issue]) -> str:
    """Flat CSV of issues, one row each, for reporting and spreadsheets.

    Deliberately one column per stored field rather than a curated subset: the
    people who ask for CSV are the ones who want to pivot it themselves.
    """
    buffer = io.StringIO()
    writer = csv.DictWriter(buffer, fieldnames=CSV_COLUMNS, extrasaction="ignore")
    writer.writeheader()
    for issue in issues:
        writer.writerow(
            {
                "id": issue.id,
                "type": issue.type,
                "identifier": issue.identifier or "",
                "severity": issue.severity or "",
                "cvss_score": _number(issue.cvss_score),
                "epss_score": _number(issue.epss_score),
                "is_kev": "true" if issue.is_kev else "false",
                "package_name": issue.package_name or "",
                "package_version": issue.package_version or "",
                "purl": issue.purl or "",
                "dependency": _dependency_label(issue.is_direct_dependency),
                "file_path": issue.file_path or "",
                "line": issue.line or "",
                "fix_state": issue.fix_state or "",
                "fix_versions": issue.fix_versions or "",
                "state": issue.state,
                "triage_status": issue.triage_status,
                "triage_justification": issue.triage_justification or "",
                "triaged_by": issue.triaged_by or "",
                "triaged_at": _isoformat(issue.triaged_at),
                "triage_expires_at": _isoformat(issue.triage_expires_at),
                "first_seen_at": _isoformat(issue.first_seen_at),
                "last_seen_at": _isoformat(issue.last_seen_at),
                "times_seen": issue.times_seen or 1,
                "link": issue.link or "",
            }
        )
    return buffer.getvalue()


def _dependency_label(is_direct: Optional[bool]) -> str:
    """Empty rather than "unknown": a column of the word "unknown" reads like a
    finding about the dependency, and the honest statement is that we have nothing
    to say."""
    if is_direct is None:
        return ""
    return "direct" if is_direct else "transitive"


def _number(value: Optional[float]) -> str:
    return "" if value is None else str(value)


def _isoformat(value) -> str:
    return value.isoformat() if value else ""

File: backend/src/domain/exports/exports.ts
import { QUALITY_TYPES } from '../gate/policy-gate';
import { canonical } from '../common/timestamp';

/**
 * Issue export formats: SARIF, OpenVEX and CSV.
 *
 * VEX is the reason triage decisions are stored in the standard's vocabulary rather than
 * as free text — this is therefore a serialization, not a translation. Every field an
 * OpenVEX statement needs is already on the issue; nothing here has to infer or invent,
 * and that is what makes the document trustworthy enough to hand to a customer or an
 * auditor.
 *
 * SARIF has another purpose: OpenVEX and CSV address people outside the pipeline, SARIF
 * exists so that a finding stops living only inside Zanshin. It is what GitHub code
 * scanning, GitLab and Azure DevOps ingest natively, and therefore what puts an issue in
 * front of the person who introduced it, annotated on the line, in the merge request —
 * instead of a dashboard they have no reason to open.
 *
 * Pure functions: the HTTP layer decides how to deliver them, and a download button in
 * the UI reuses them unchanged.
 *
 * Checked against `test/vectors/exports.json`, produced by the real Python module.
 */

export const OPENVEX_CONTEXT = 'https://openvex.dev/ns/v0.2.0';
export const SARIF_VERSION = '2.1.0';
export const SARIF_SCHEMA = 'https://json.schemastore.org/sarif-2.1.0.json';

const STATE_RESOLVED = 'resolved';
const TRIAGE_UNDER_REVIEW = 'under_review';
const TRIAGE_AFFECTED = 'affected';
const TRIAGE_NOT_AFFECTED = 'not_affected';
const TRIAGE_FIXED = 'fixed';

/**
 * Zanshin's triage vocabulary is already OpenVEX's, with one exception: `under_review` is
 * spelled `under_investigation` in the specification.
 */
const VEX_STATUS: Record<string, string> = {
    [TRIAGE_UNDER_REVIEW]: 'under_investigation',
    [TRIAGE_AFFECTED]: 'affected',
    [TRIAGE_NOT_AFFECTED]: 'not_affected',
    [TRIAGE_FIXED]: 'fixed'
};

/**
 * SARIF has four levels and no notion of "critical". Anything a security tool would call
 * critical or high has to land on `error`, because `warning` is what a reviewer scrolls
 * past without reading.
 */
const SARIF_LEVEL: Record<string, string> = {
    critical: 'error',
    high: 'error',
    medium: 'warning',
    low: 'note',
    negligible: 'note',
    unknown: 'warning'
};

/**
 * GitHub sorts and filters on this property, **not** on `level`: it is what keeps a
 * critical distinguishable from a high once both are `error`. The values follow the CVSS
 * bands GitHub documents.
 */
const SECURITY_SEVERITY: Record<string, string> = {
    critical: '9.5',
    high: '8.0',
    medium: '5.5',
    low: '3.0',
    negligible: '1.0'
};

const ISSUE_TYPE_LABEL: Record<string, string> = {
    vulnerability: 'Vulnerability',
    secret: 'Exposed secret',
    iac: 'Infrastructure configuration',
    license: 'License',
    eol: 'End of life',
    ai_review: 'AI review',
    sast: 'Vulnerable code',
    quality: 'Code quality'
};

export const CSV_COLUMNS = [
    'id',
    'type',
    'identifier',
    'severity',
    'cvss_score',
    'epss_score',
    'is_kev',
    'package_name',
    'package_version',
    'purl',
    'dependency',
    'file_path',
    'line',
    'fix_state',
    'fix_versions',
    'state',
    'triage_status',
    'triage_justification',
    'triaged_by',
    'triaged_at',
    'triage_expires_at',
    'first_seen_at',
    'last_seen_at',
    'times_seen',
    'link'
] as const;

/**
 * An issue as the exports read it.
 *
 * Timestamps are canonicalized on the way out — never rendered by the driver's own
 * formatting, which loses the microsecond and applies a timezone (see
 * `common/timestamp.ts`).
 */
export interface ExportableIssue {
    id: number;
    fingerprint: string | null;
    type: string | null;
    identifier: string | null;
    severity: string | null;
    cvssScore: number | null;
    epssScore: number | null;
    isKev: boolean | null;
    packageName: string | null;
    packageVersion: string | null;
    purl: string | null;
    isDirectDependency: boolean | null;
    filePath: string | null;
    line: number | null;
    fixState: string | null;
    fixVersions: string | null;
    link: string | null;
    description: string | null;
    state: string | null;
    triageStatus: string | null;
    triageJustification: string | null;
    triageComment: string | null;
    triagedBy: string | null;
    triagedAt: Date | null;
    triageExpiresAt: Date | null;
    firstSeenAt: Date | null;
    lastSeenAt: Date | null;
    timesSeen: number | null;
}

// --------------------------------------------------------------------------- OpenVEX

export interface OpenVexOptions {
    author: string;
    productId: string;
    documentId: string;
    /** Supplied by the caller: a VEX document is an assertion about who said what and
     *  when, which belongs to whoever publishes it, not to a utility function. */
    timestamp: Date;
    version?: number;
}

/**
 * An OpenVEX document for a product, from its vulnerability issues.
 *
 * Only `vulnerability` issues are included: VEX is defined over vulnerability
 * identifiers, and a hardcoded secret or a failed IaC check has no CVE to make a
 * statement about. Issues with no identifier are discarded for the same reason — an
 * anonymous statement is not one.
 */
export function buildOpenVexDocument(issues: Iterable<ExportableIssue>, options: OpenVexOptions): Record<string, unknown> {
    const statements: Record<string, unknown>[] = [];

    for (const issue of issues) {
        if (issue.type !== 'vulnerability' || !issue.identifier) continue;

        let status = VEX_STATUS[issue.triageStatus ?? ''] ?? 'under_investigation';
        // An issue that is resolved and was never triaged is factually fixed: the
        // scanner stopped seeing it. Saying "under investigation" about something that has
        // gone would be misleading in a document made to answer exactly that.
        if (issue.state === STATE_RESOLVED && issue.triageStatus === TRIAGE_UNDER_REVIEW) {
            status = 'fixed';
        }

        const statement: Record<string, unknown> = {
            vulnerability: { name: issue.identifier },
            products: [{ '@id': options.productId }],
            status
        };

        if (status === 'not_affected') {
            // Required by the specification for this status, and guaranteed present by
            // the triage service.
            statement.justification = issue.triageJustification;
            if (issue.triageComment) statement.impact_statement = issue.triageComment;
        } else if (status === 'affected' && issue.triageComment) {
            // For "affected", the free text belongs to the action statement.
            statement.action_statement = issue.triageComment;
        }

        if (issue.purl) {
            statement.products = [{ '@id': options.productId, identifiers: { purl: issue.purl } }];
        }
        // RFC 3339, as the OpenVEX specification requires. The previous document carried
        // "2026-08-10T08:00:00" with no timezone, which is not a valid instant under that
        // standard: a strict consumer was entitled to refuse it.
        if (issue.triagedAt) statement.timestamp = canonical(issue.triagedAt);
        else if (issue.lastSeenAt) statement.timestamp = canonical(issue.lastSeenAt);

        statements.push(statement);
    }

    return {
        '@context': OPENVEX_CONTEXT,
        '@id': options.documentId,
        author: options.author,
        timestamp: canonical(options.timestamp),
        version: options.version ?? 1,
        tooling: 'Zanshin',
        statements
    };
}

// ----------------------------------------------------------------------------- SARIF

export interface SarifOptions {
    targetName: string;
    toolVersion?: string;
    informationUri?: string | null;
}

/**
 * A SARIF 2.1.0 log for a target's issues.
 *
 * Decisions worth stating, because SARIF is permissive enough that a technically valid
 * document can still be useless in a code scanning interface:
 *
 * - **Triaged issues are `suppressions`, not omissions.** Removing them would make a
 *   platform report them as new on the next upload, undoing the triage work; and a
 *   suppression carries its justification, so the reviewer sees *why* it is set aside.
 *   `not_affected` and `fixed` are suppressed, `affected` is not — deciding that an issue
 *   is real must stay visible.
 * - **Resolved issues are excluded.** They are gone; SARIF's job is the current state of
 *   the branch being built.
 * - **`partialFingerprints` carries Zanshin's fingerprint**, which lets the platform
 *   match an issue from one upload to the next even if the file moves or the line shifts.
 * - **Every result has a location**, falling back to the repository root when a
 *   dependency issue has no file. GitHub silently discards results with no location, so
 *   an "honestly empty" location would make the vulnerability findings disappear — that
 *   is, most of them.
 *
 * Rules are emitted per distinct identifier rather than per issue: that is what the SARIF
 * model means by a rule, and what lets a platform group them.
 */
export function buildSarifDocument(issues: Iterable<ExportableIssue>, options: SarifOptions): Record<string, unknown> {
    const current = [...issues].filter((issue) => issue.state !== STATE_RESOLVED);

    // A `Map` and not an object: the rules' insertion order determines `ruleIndex`, and
    // an object would reorder keys that look like integers.
    const rules = new Map<string, Record<string, unknown>>();
    const ruleIndex = new Map<string, number>();
    const results: Record<string, unknown>[] = [];

    for (const issue of current) {
        const ruleId = sarifRuleId(issue);
        if (!rules.has(ruleId)) {
            ruleIndex.set(ruleId, rules.size);
            rules.set(ruleId, sarifRule(issue, ruleId));
        }

        const properties: Record<string, unknown> = {
            zanshinIssueId: issue.id,
            type: issue.type,
            firstSeen: issue.firstSeenAt ? canonical(issue.firstSeenAt) : '',
            timesSeen: issue.timesSeen || 1
        };
        if (issue.isDirectDependency !== null && issue.isDirectDependency !== undefined) {
            properties.dependency = issue.isDirectDependency ? 'direct' : 'transitive';
        }

        const result: Record<string, unknown> = {
            ruleId,
            ruleIndex: ruleIndex.get(ruleId),
            level: SARIF_LEVEL[(issue.severity || 'unknown').toLowerCase()] ?? 'warning',
            message: { text: sarifMessage(issue) },
            locations: [sarifLocation(issue)],
            partialFingerprints: { zanshinIssueFingerprint: issue.fingerprint },
            properties
        };

        if (isSuppressed(issue)) {
            result.suppressions = [
                {
                    // "external": the decision was taken in Zanshin, not in a source
                    // annotation, which is what this kind of suppression documents.
                    kind: 'external',
                    justification: suppressionJustification(issue)
                }
            ];
        }
        results.push(result);
    }

    const driver: Record<string, unknown> = {
        name: 'Zanshin',
        version: options.toolVersion ?? '1.0.0'
    };
    // Absent rather than null when not supplied: `**({...} if x else {})`.
    if (options.informationUri) driver.informationUri = options.informationUri;
    driver.rules = [...rules.values()];

    return {
        $schema: SARIF_SCHEMA,
        version: SARIF_VERSION,
        runs: [
            {
                tool: { driver },
                results,
                properties: { target: options.targetName }
            }
        ]
    };
}

/** `security` only for what is genuinely a security finding. */
function sarifTags(issue: ExportableIssue): string[] {
    if (issue.type != null && QUALITY_TYPES.includes(issue.type)) return ['quality', issue.type];
    return ['security', issue.type as string];
}

/**
 * Stable, and partitioned by type.
 *
 * A gitleaks rule and a checkov check can collide on an identifier, and a platform
 * indexed on `ruleId` would then merge two unrelated classes of issue under one title.
 */
function sarifRuleId(issue: ExportableIssue): string {
    return `zanshin/${issue.type}/${issue.identifier || 'unspecified'}`;
}

function sarifRule(issue: ExportableIssue, ruleId: string): Record<string, unknown> {
    const label = (issue.type != null ? ISSUE_TYPE_LABEL[issue.type] : undefined) ?? issue.type;
    const properties: Record<string, unknown> = { tags: sarifTags(issue) };

    const rule: Record<string, unknown> = {
        id: ruleId,
        name: (issue.identifier || issue.type || '').replace(/ /g, ''),
        shortDescription: { text: `${label}: ${issue.identifier || 'unidentified'}` },
        properties
    };

    if (issue.description) rule.fullDescription = { text: issue.description.slice(0, 1000) };
    if (issue.link) rule.helpUri = issue.link;

    const severity = (issue.severity || '').toLowerCase();
    if (severity in SECURITY_SEVERITY) properties['security-severity'] = SECURITY_SEVERITY[severity];

    return rule;
}

/**
 * What the developer reads in the merge request, hence what says what to do.
 *
 * The fixed version is the most useful thing to put in front of someone who has thirty
 * seconds: it turns "there is a CVE" into "change this line".
 */
function sarifMessage(issue: ExportableIssue): string {
    const parts: string[] = [];
    if (issue.packageName) {
        parts.push(issue.packageVersion ? `${issue.packageName} ${issue.packageVersion}` : issue.packageName);
    }
    parts.push(issue.identifier || (issue.type != null ? ISSUE_TYPE_LABEL[issue.type] : undefined) || (issue.type as string));

    let message = parts.join(' — ');
    if (issue.fixVersions) message += ` — fixed in ${issue.fixVersions}`;
    else if (issue.fixState === 'not-fixed') message += ' — no published fix';
    if (issue.isKev) message += ' — known active exploitation (CISA KEV)';
    if (issue.isDirectDependency === false) message += ' — transitive dependency';
    return message;
}

function sarifLocation(issue: ExportableIssue): Record<string, unknown> {
    const physicalLocation: Record<string, unknown> = {
        // A relative URI, as SARIF requires for source the consumer resolves against the
        // repository it has just checked out.
        artifactLocation: { uri: issue.filePath || '.' }
    };
    if (issue.line) physicalLocation.region = { startLine: Math.trunc(issue.line) };

    const location: Record<string, unknown> = { physicalLocation };
    if (issue.purl) location.logicalLocations = [{ name: issue.purl, kind: 'package' }];
    return location;
}

function isSuppressed(issue: ExportableIssue): boolean {
    return issue.triageStatus === TRIAGE_NOT_AFFECTED || issue.triageStatus === TRIAGE_FIXED;
}

function suppressionJustification(issue: ExportableIssue): string {
    const parts: string[] = [issue.triageStatus as string];
    if (issue.triageJustification) parts.push(issue.triageJustification);
    if (issue.triageComment) parts.push(issue.triageComment);
    if (issue.triagedBy) parts.push(`decided by ${issue.triagedBy}`);
    // `.date().isoformat()` in Python: the date part alone, hence the first ten
    // characters of an isoformat.
    if (issue.triageExpiresAt) parts.push(`to review on ${issue.triageExpiresAt.toISOString().slice(0, 10)}`);
    return parts.join(' — ');
}

// ------------------------------------------------------------------------------- CSV

/**
 * Flat CSV of the issues, one row each, for reporting and spreadsheets.
 *
 * Deliberately one column per stored field rather than a chosen subset: the people who
 * ask for CSV are the ones who want to cross-reference it themselves.
 */
export function buildIssuesCsv(issues: Iterable<ExportableIssue>): string {
    const rows: string[] = [CSV_COLUMNS.join(',')];

    for (const issue of issues) {
        rows.push(
            [
                String(issue.id),
                issue.type ?? '',
                issue.identifier || '',
                issue.severity || '',
                pythonNumber(issue.cvssScore),
                pythonNumber(issue.epssScore),
                issue.isKev ? 'true' : 'false',
                issue.packageName || '',
                issue.packageVersion || '',
                issue.purl || '',
                dependencyLabel(issue.isDirectDependency),
                issue.filePath || '',
                issue.line ? String(issue.line) : '',
                issue.fixState || '',
                issue.fixVersions || '',
                issue.state ?? '',
                issue.triageStatus ?? '',
                issue.triageJustification || '',
                issue.triagedBy || '',
                // Canonicalized as everywhere else: a CSV compared byte for byte must not
                // depend on the timezone of the machine that produced it.
                issue.triagedAt ? canonical(issue.triagedAt) : '',
                issue.triageExpiresAt ? canonical(issue.triageExpiresAt) : '',
                issue.firstSeenAt ? canonical(issue.firstSeenAt) : '',
                issue.lastSeenAt ? canonical(issue.lastSeenAt) : '',
                String(issue.timesSeen || 1),
                issue.link || ''
            ]
                .map(quoteCsvField)
                .join(',')
        );
    }

    // Python's `csv.writer` terminates its rows with CRLF (the "excel" dialect),
    // including the last one. Using `\n` would produce a file most tools would read
    // anyway — hence a divergence nobody would notice until a strict consumer refused
    // it.
    return rows.map((row) => `${row}\r\n`).join('');
}

/**
 * The characters by which a spreadsheet decides a cell is a **formula**.
 *
 * Excel, LibreOffice and Google Sheets evaluate a cell starting with any of them. Tab and
 * carriage return are in the list because Excel skips them before resuming its parse:
 * `\t=cmd|…` is evaluated as `=cmd|…`.
 */
const FORMULA_PREFIX = /^[=+\-@\t\r]/;

/**
 * Quoting in the manner of `QUOTE_MINIMAL`, **preceded by formula neutralization**.
 *
 * **This file's content comes from scanned repositories**, hence from outside the trust
 * boundary: a package name, a file path, a rule identifier are chosen by whoever can
 * commit to the target. The reader, meanwhile, is a security operator opening the file in
 * a spreadsheet — which is the whole point of the CSV export.
 *
 * A package named `=cmd|'/c calc'!A1` executes on open; `=HYPERLINK(...&A1&B1)` exfiltrates
 * the neighbouring cells — that is, the rest of the backlog — to a host of the attacker's
 * choosing, with no prompt at all. The apostrophe forces text mode.
 *
 * **Quoting does not protect**: the spreadsheet strips the quotes before evaluating.
 * Neutralization must therefore precede quoting, not stand in for it.
 */
function quoteCsvField(value: string): string {
    const safe = FORMULA_PREFIX.test(value) ? `'${value}` : value;
    if (!/[",\r\n]/.test(safe)) return safe;
    return `"${safe.replace(/"/g, '""')}"`;
}

/**
 * Empty rather than "unknown": a column filled with the word "unknown" reads as a finding
 * about the dependency, when the honest statement is that we have nothing to say about it.
 */
function dependencyLabel(isDirect: boolean | null | undefined): string {
    if (isDirect === null || isDirect === undefined) return '';
    return isDirect ? 'direct' : 'transitive';
}

/**
 * Python's `str(float)`, which keeps the decimal on whole values: `str(9.0)` returns
 * "9.0", where `String(9)` returns "9" in JavaScript.
 *
 * On a CVSS score column, half the values are whole numbers — the divergence would
 * therefore touch every other row of an export handed to an auditor.
 */
function pythonNumber(value: number | null | undefined): string {
    if (value === null || value === undefined) return '';
    return Number.isInteger(value) ? `${value}.0` : String(value);
}
